File: day_20_plaintext_ciphertext_keys_encryption_and_decryption/src/encryption_demo.py
"""
Day 20: Plaintext, Ciphertext, Keys, Encryption and Decryption

Main command-line demonstration for the learning project.

The program shows the complete lifecycle:

    Plaintext
        |
        | encryption + secret key
        v
    Ciphertext
        |
        | decryption + same secret key
        v
    Plaintext

The implementation uses Fernet from the Python cryptography package.
It does not implement cryptographic algorithms from scratch.
"""

from __future__ import annotations

import argparse
import sys

from cryptography.fernet import Fernet, InvalidToken


def generate_key() -> bytes:
    """Generate a new symmetric encryption key."""
    return Fernet.generate_key()


def encrypt(plaintext: str, key: bytes) -> str:
    """Encrypt plaintext using the supplied Fernet key."""
    if not plaintext:
        raise ValueError("Plaintext cannot be empty.")

    return Fernet(key).encrypt(
        plaintext.encode("utf-8")
    ).decode("utf-8")


def decrypt(ciphertext: str, key: bytes) -> str:
    """Decrypt ciphertext using the supplied Fernet key."""
    if not ciphertext:
        raise ValueError("Ciphertext cannot be empty.")

    try:
        return Fernet(key).decrypt(
            ciphertext.encode("utf-8")
        ).decode("utf-8")
    except InvalidToken as exc:
        raise ValueError(
            "Decryption failed. Check the encryption key and "
            "make sure the ciphertext has not been modified."
        ) from exc


def print_separator() -> None:
    """Print a visual separator."""
    print("=" * 72)


def demonstrate(message: str) -> None:
    """Run and display a complete encryption/decryption demonstration."""
    print_separator()
    print("DAY 20 - PLAINTEXT, CIPHERTEXT, KEYS,")
    print("         ENCRYPTION AND DECRYPTION")
    print_separator()
    print()

    print("1. Original data")
    print("-" * 72)
    print(f"Plaintext:\n{message}")
    print()

    key = generate_key()

    print("2. Generated encryption key")
    print("-" * 72)
    print(key.decode("utf-8"))
    print()
    print(
        "The key is sensitive information. Anyone who obtains the key "
        "may be able to decrypt data protected by it."
    )
    print()

    ciphertext = encrypt(message, key)

    print("3. Encryption")
    print("-" * 72)
    print("The plaintext is transformed into ciphertext.")
    print()
    print(f"Ciphertext:\n{ciphertext}")
    print()

    recovered = decrypt(ciphertext, key)

    print("4. Decryption")
    print("-" * 72)
    print("The ciphertext is transformed back into plaintext.")
    print()
    print(f"Recovered plaintext:\n{recovered}")
    print()

    print("5. Verification")
    print("-" * 72)

    if recovered == message:
        print("SUCCESS: decrypted text matches the original plaintext.")
    else:
        print("ERROR: decrypted text does not match the original plaintext.")

    print()
    print_separator()


def encrypt_mode(message: str) -> None:
    """Generate a key and encrypt a supplied message."""
    key = generate_key()
    ciphertext = encrypt(message, key)

    print_separator()
    print("ENCRYPTION RESULT")
    print_separator()
    print(f"Plaintext : {message}")
    print(f"Key       : {key.decode('utf-8')}")
    print(f"Ciphertext: {ciphertext}")
    print_separator()


def decrypt_mode(ciphertext: str, key_text: str) -> None:
    """Decrypt supplied ciphertext using a supplied key."""
    try:
        key = key_text.encode("utf-8")
        plaintext = decrypt(ciphertext, key)
    except (ValueError, TypeError) as exc:
        print(f"ERROR: {exc}")
        sys.exit(1)

    print_separator()
    print("DECRYPTION RESULT")
    print_separator()
    print(f"Ciphertext: {ciphertext}")
    print(f"Plaintext : {plaintext}")
    print_separator()


def build_parser() -> argparse.ArgumentParser:
    """Create the command-line argument parser."""
    parser = argparse.ArgumentParser(
        description=(
            "Educational demonstration of plaintext, ciphertext, "
            "keys, encryption and decryption."
        )
    )

    parser.add_argument(
        "--message",
        default=(
            "Cryptography protects information by transforming "
            "readable data into protected ciphertext."
        ),
        help="Plaintext message used by the demonstration.",
    )

    parser.add_argument(
        "--encrypt",
        metavar="TEXT",
        help="Encrypt TEXT and print the generated key and ciphertext.",
    )

    parser.add_argument(
        "--decrypt",
        metavar="CIPHERTEXT",
        help="Decrypt CIPHERTEXT using --key.",
    )

    parser.add_argument(
        "--key",
        help="Fernet key used with --decrypt.",
    )

    return parser


def main() -> None:
    """Application entry point."""
    parser = build_parser()
    args = parser.parse_args()

    if args.encrypt is not None:
        encrypt_mode(args.encrypt)
        return

    if args.decrypt is not None:
        if not args.key:
            parser.error("--decrypt requires --key")

        decrypt_mode(args.decrypt, args.key)
        return

    demonstrate(args.message)


if __name__ == "__main__":
    main()

File: day_20_plaintext_ciphertext_keys_encryption_and_decryption/cpp/encryption_demo.cpp
/*
    Day 20: Plaintext, Ciphertext, Keys, Encryption and Decryption

    Educational C++ demonstration of the conceptual encryption
    and decryption workflow.

    Important:
    This program demonstrates the relationship between plaintext,
    ciphertext, keys, encryption, and decryption. It does NOT
    implement a production cryptographic algorithm.

    Production applications should use established cryptographic
    libraries rather than custom cryptographic implementations.
*/

#include <iomanip>
#include <iostream>
#include <stdexcept>
#include <sstream>
#include <string>
#include <vector>

using namespace std;

/*
    Educational XOR transformation.

    XOR is useful for demonstrating the basic idea that a key can
    participate in a transformation, but this is NOT secure
    encryption and must never be used to protect real information.
*/
string xorTransform(const string& input, const string& key)
{
    if (key.empty())
    {
        throw invalid_argument("Key cannot be empty.");
    }

    string output = input;

    for (size_t i = 0; i < input.size(); ++i)
    {
        output[i] = input[i] ^ key[i % key.size()];
    }

    return output;
}

/*
    Convert binary data into hexadecimal text so that the
    educational ciphertext can be displayed safely.
*/
string toHex(const string& data)
{
    ostringstream output;

    for (unsigned char character : data)
    {
        output << hex
               << setw(2)
               << setfill('0')
               << static_cast<int>(character);
    }

    return output.str();
}

/*
    Convert hexadecimal text back into binary data.
*/
string fromHex(const string& hexData)
{
    if (hexData.size() % 2 != 0)
    {
        throw invalid_argument(
            "Hexadecimal ciphertext must contain an even number "
            "of characters."
        );
    }

    string result;

    for (size_t i = 0; i < hexData.size(); i += 2)
    {
        unsigned int value;

        string byteString = hexData.substr(i, 2);

        stringstream stream;
        stream << hex << byteString;
        stream >> value;

        result.push_back(
            static_cast<char>(value)
        );
    }

    return result;
}

void printSeparator()
{
    cout << string(72, '=') << '\n';
}

int main()
{
    try
    {
        const string plaintext =
            "Plaintext is readable information.";

        const string key =
            "DAY20KEY";

        printSeparator();

        cout << "DAY 20 - PLAINTEXT, CIPHERTEXT, KEYS,\n";
        cout << "         ENCRYPTION AND DECRYPTION\n";

        printSeparator();

        cout << "\n1. PLAINTEXT\n";
        cout << "------------------------------------------------------------------------\n";
        cout << plaintext << '\n';

        cout << "\n2. ENCRYPTION KEY\n";
        cout << "------------------------------------------------------------------------\n";
        cout << key << '\n';

        /*
            Educational transformation.

            XOR is reversible:

                plaintext XOR key = ciphertext

                ciphertext XOR key = plaintext

            This property demonstrates the relationship between
            encryption and decryption but does NOT make XOR itself
            a secure encryption system.
        */
        const string binaryCiphertext =
            xorTransform(plaintext, key);

        const string ciphertext =
            toHex(binaryCiphertext);

        cout << "\n3. CIPHERTEXT\n";
        cout << "------------------------------------------------------------------------\n";
        cout << ciphertext << '\n';

        cout << "\n4. DECRYPTION\n";
        cout << "------------------------------------------------------------------------\n";

        const string recoveredBinary =
            fromHex(ciphertext);

        const string recoveredPlaintext =
            xorTransform(recoveredBinary, key);

        cout << recoveredPlaintext << '\n';

        cout << "\n5. VERIFICATION\n";
        cout << "------------------------------------------------------------------------\n";

        if (recoveredPlaintext == plaintext)
        {
            cout << "SUCCESS: decrypted text matches the original plaintext.\n";
        }
        else
        {
            cout << "ERROR: decrypted text does not match the original.\n";
            return 1;
        }

        cout << "\n6. IMPORTANT SECURITY NOTE\n";
        cout << "------------------------------------------------------------------------\n";
        cout << "This XOR implementation is for education only.\n";
        cout << "It is NOT secure cryptography.\n";
        cout << "Real applications should use established cryptographic libraries.\n";

        cout << '\n';
        printSeparator();

        return 0;
    }
    catch (const exception& error)
    {
        cerr << "ERROR: " << error.what() << '\n';
        return 1;
    }
}